types/datetime.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	if !(unicode.IsDigit(rune(s[14])) &&
+		unicode.IsDigit(rune(s[15]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid minute", ErrDatetime)
+	}
+	minute = 10*int(rune(s[14])-'0') + int(rune(s[15])-'0')
+
+	if s[16] != ':' {
+		return Datetime{}, fmt.Errorf("%w: unexpected character %s", ErrDatetime, strconv.QuoteRune(rune(s[16])))
+	}
+
+	if !(unicode.IsDigit(rune(s[17])) &&
+		unicode.IsDigit(rune(s[18]))) {
+		return Datetime{}, fmt.Errorf("%w: invalid second", ErrDatetime)
+	}
+	second = 10*int(rune(s[17])-'0') + int(rune(s[18])-'0')
+
+	// At this point, things are variable.
+	// 19 can be ., in which case we have milliseconds. (SSS)
+	//   ... but we'll still need a Z, or offset. So, we'll introduce
+	//       trailerOffset to account for where this starts.
+	trailerOffset := 19
+	if s[19] == '.' {
+		if length < 23 {
+			return Datetime{}, fmt.Errorf("%w: invalid millisecond", ErrDatetime)
+		}
+
+		if !(unicode.IsDigit(rune(s[20])) &&
+			unicode.IsDigit(rune(s[21])) &&
+			unicode.IsDigit(rune(s[22]))) {
+			return Datetime{}, fmt.Errorf("%w: invalid millisecond", ErrDatetime)
+		}
+
+		milli = 100*int(rune(s[20])-'0') + 10*int(rune(s[21])-'0') + int(rune(s[22])-'0')
+		trailerOffset = 23
+	}
+
+	if length == trailerOffset {
+		return Datetime{}, fmt.Errorf("%w: expected time zone designator", ErrDatetime)
+	}
+
+	// At this point, we can only have 2 possible lengths. Anything else is an error.
+	switch s[trailerOffset] {
+	case 'Z':
+		if length > trailerOffset+1 {
+			// If something comes after the Z, it's an error
+			return Datetime{}, fmt.Errorf("%w: unexpected trailer after time zone designator", ErrDatetime)
+		}
+	case '+', '-':
+		sign := 1
+		if s[trailerOffset] == '-' {
+			sign = -1
+		}
+
+		if length > trailerOffset+5 {
+			return Datetime{}, fmt.Errorf("%w: unexpected trailer after time zone designator", ErrDatetime)
+		} else if length != trailerOffset+5 {
+			return Datetime{}, fmt.Errorf("%w: invalid time zone offset", ErrDatetime)
+		}
+
+		// get the time zone offset hhmm.
+		if !(unicode.IsDigit(rune(s[trailerOffset+1])) &&
+			unicode.IsDigit(rune(s[trailerOffset+2])) &&
+			unicode.IsDigit(rune(s[trailerOffset+3])) &&
+			unicode.IsDigit(rune(s[trailerOffset+4]))) {
+			return Datetime{}, fmt.Errorf("%w: invalid time zone offset", ErrDatetime)
+		}
+
+		hh := time.Duration(10*int64(rune(s[trailerOffset+1])-'0')+int64(rune(s[trailerOffset+2])-'0')) * time.Hour
+		mm := time.Duration(10*int64(rune(s[trailerOffset+3])-'0')+int64(rune(s[trailerOffset+4])-'0')) * time.Minute
+		offset = time.Duration(sign) * (hh + mm)
+
+	default:
+		return Datetime{}, fmt.Errorf("%w: invalid time zone designator", ErrDatetime)
+	}
+
+	t := time.Date(year, time.Month(month), day,
+		hour, minute, second,
+		int(time.Duration(milli)*time.Millisecond), time.UTC)
+	t = t.Add(offset)
+	return Datetime{Value: t.UnixMilli()}, nil
+}
+
+// Equal returns true if the lfs represents the same timestamp as the rhs.
```

Comment:
I think "lhs" instead of "lfs"

I feel like "lhs" and "rhs" is a bit odd.  Maybe have it phrased "Equal returns true if the input represents the same timestamp."